In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
% matplotlib inline
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import robust_scale
from sklearn import preprocessing
from sklearn.preprocessing import OneHotEncoder
from sklearn import metrics
from sklearn.model_selection import train_test_split
from sklearn import datasets, linear_model
import scipy.stats as stats
import statistics

In [2]:
df=pd.read_csv('worker_matrix.csv')

#### If we calculate correlation vs work duration for individual, it comes as follows

In [3]:
df.sort_values(by='No of tasks',ascending = False)

Worker Number                 Worker_id  No of tasks  Correlation
67             67  54896d76fc4e7e0f300000b4          659    -0.011211
51             51  523949ca3465040cdb0078ac          384    -0.076515
16             16  512dbf09c32206698e000518          271     0.438533
13             13  5119d07bed964935a800008b          244     0.055305
2               2  50531712cb022432f3000f27          195    -0.085539
1               1  503b89aacb02245d320001a4          136     0.063627
87             87  5678e1a8fc4e7e9e0b000086          133     0.045923
33             33  51e60f42863a99e7d5025708          124    -0.163825
49             49  5232d065863a99092101733d          121     0.023250
72             72  566fc9c6fc4e7e1195000d55          114    -0.020008
0               0  501767d1cb022453170004dc          102     0.103216
55             55  525e0c13807fcd982f03e0a0          101     0.018961
81             81  5678c14efc4e7e8c1500003d           92     0.294536
38             38  5208c83db684b001ec000c79           91     0.145620
69             69  5668f692fc4e7e2ef60007c0           87     0.172263
60             60  52833d27863a9990e702e220           86    -0.134650
4               4  50b59592ed964941970005b6           85     0.123976
5               5  50b6078eed9649419f000758           81    -0.148474
41             41  521f2fa7863a99cb29000159           81     0.121319
71             71  566fc951fc4e7e3aac0008a8           80     0.321067
62             62  52ae981f863a9953ae064312           78     0.076988
39             39  521ae408b684b0370203c22e           72    -0.121300
56             56  526f7299346504becc043687           69     0.010592
36             36  51ff1ed1863a99a5d3002c40           69    -0.029888
65             65  52dc8819b684b02135020212           69     0.106003
3               3  5059b9b4cb02246c1200017d           63    -0.329779
82             82  5678c86cfc4e7e6e7e00004c           62    -0.020661
8               8  50ca9696cb0224456b0012ba           59    -0.266461
70             70  566fc7ddfc4e7e7cd7000cff           58     0.315660
25             25  518ca54f863a9902a40010db           57     0.326403
..            ...                       ...          ...          ...
59             59  527484feb684b0734e021055           12     0.045328
45             45  5229b417b684b08fa8006f8b           12     0.386859
9               9  50cdd243ed9649443f00035b           11    -0.597417
79             79  5677cd34fc4e7e267400002c           11    -0.455873
23             23  5178ebb6ed9649765a000166           11     0.225737
63             63  52bad53e863a999598061739           10    -0.457106
74             74  56711d03fc4e7edcd20009f4           10     0.399174
92             92  573c2410651bb1ae90000516            9     0.306417
28             28  51a33eb1863a99725c00396c            9     0.021536
90             90  570b4757cb63b02e36000457            8     0.335827
73             73  56711cfafc4e7edcd20009e6            8    -0.260458
61             61  52a6f377863a99a06f001cc4            8     0.408486
43             43  522847d0b684b0acd8003551            7     0.216052
11             11  51089b31c322063455000002            6    -0.146566
68             68  54d95e9ffc4e7e009100281d            6     0.117468
84             84  5678c87cfc4e7e6e7e000068            4     0.994158
42             42  5227f4bc863a996ea200ea7a            4     0.250835
44             44  52286b2eb684b05f3f00616e            4     0.004472
19             19  513ff1fac3220622d8002e1c            4     0.136509
35             35  51e65c79b684b072b802aa05            2     1.000000
32             32  51c922b8863a99745f013211            2    -1.000000
40             40  521ae474b684b0a1a903cb63            2     1.000000
31             31  51c7f982b684b0ab8f026427            2    -1.000000
58             58  527345fb34650437bd011f58            1          NaN
14             14  512c57a2ed96497cdf0015e0            1          NaN
12            

> The encouraginf side is, for the workers with maximum number of tasks, correlation is negative. That means accuracy decreases as work duration increases. 

## Converting all the accuracy values in binary 

In [4]:
worker =pd.read_csv('worker.csv')

In [5]:
worker.head()

Unnamed: 0                   task_id                 worker_id  \
0           0  57a2997ee2013f0c5a5d9f33  501767d1cb022453170004dc   
1           1  57a2997ee2013f0c5a5d9f34  501767d1cb022453170004dc   
2           2  57a2997ee2013f0c5a5d9f35  501767d1cb022453170004dc   
3           3  57a2997ee2013f0c5a5d9f36  501767d1cb022453170004dc   
4           4  57a2997ee2013f0c5a5d9f37  501767d1cb022453170004dc   

   work_duration  accuracy  mousemove_distance  mouseclick_button  
0         88.318  0.833333           19.722592           0.068966  
1         34.392  1.000000            7.173393           0.376812  
2         18.640  1.000000           22.958495           0.061224  
3         15.841  0.833333            8.415718           0.093023  
4         15.844  1.000000            2.868044           0.155556

In [6]:
worker = pd.DataFrame(worker[['task_id','worker_id','work_duration','accuracy','mousemove_distance','mouseclick_button']])

In [7]:
worker['accuracy'] = worker['accuracy'] * 6

In [8]:
worker['accuracy'] = worker['accuracy'].round()

In [9]:
worker['accuracy'].value_counts()

6.0    2626
5.0     923
3.0     476
0.0     475
4.0     227
1.0     218
2.0      55
Name: accuracy, dtype: int64

In [10]:
worker.loc[worker['accuracy']!=6, 'accuracy'] = 0

In [11]:
worker['accuracy'].value_counts()

6.0    2626
0.0    2374
Name: accuracy, dtype: int64

In [12]:
worker

task_id                 worker_id  work_duration  \
0     57a2997ee2013f0c5a5d9f33  501767d1cb022453170004dc         88.318   
1     57a2997ee2013f0c5a5d9f34  501767d1cb022453170004dc         34.392   
2     57a2997ee2013f0c5a5d9f35  501767d1cb022453170004dc         18.640   
3     57a2997ee2013f0c5a5d9f36  501767d1cb022453170004dc         15.841   
4     57a2997ee2013f0c5a5d9f37  501767d1cb022453170004dc         15.844   
5     57a2997ee2013f0c5a5d9f38  501767d1cb022453170004dc          9.907   
6     57a2997ee2013f0c5a5d9f39  501767d1cb022453170004dc         18.598   
7     57a2997ee2013f0c5a5d9f3a  501767d1cb022453170004dc         75.675   
8     57a2997ee2013f0c5a5d9f3b  501767d1cb022453170004dc         17.629   
9     57a2997ee2013f0c5a5d9f3c  501767d1cb022453170004dc        114.393   
10    57a2997ee2013f0c5a5d9f3d  501767d1cb022453170004dc         17.029   
11    57a2997ee2013f0c5a5d9f3e  501767d1cb022453170004dc        111.812   
12    57a29b3de2013f6e4adcca36  501767d1cb022453170004dc         28.240   
13    57a29b3de2013f6e4adcca37  501767d1cb022453170004dc         26.275   
14    57a29b3de2013f6e4adcca38  501767d1cb022453170004dc         50.592   
15    57a29b3de2013f6e4adcca39  501767d1cb022453170004dc         18.015   
16    57a29b3de2013f6e4adcca3a  501767d1cb022453170004dc         19.247   
17    57a29b3de2013f6e4adcca3b  501767d1cb022453170004dc         27.256   
18    57a29b3de2013f6e4adcca3c  501767d1cb022453170004dc         15.187   
19    57a29b3de2013f6e4adcca3d  501767d1cb022453170004dc         15.498   
20    57a29b3de2013f6e4adcca3e  501767d1cb022453170004dc          9.876   
21    57a29b3de2013f6e4adcca3f  501767d1cb022453170004dc        130.237   
22    57a29b3de2013f6e4adcca40  501767d1cb022453170004dc         31.755   
23    57a29b3de2013f6e4adcca41  501767d1cb022453170004dc         18.046   
24    57a29d44e2013f0c5a5da1dc  501767d1cb022453170004dc         38.389   
25    57a29d44e2013f0c5a5da1dd  501767d1cb022453170004dc          5.014   
26    57a29d44e2013f0c5a5da1de  501767d1cb022453170004dc         33.238   
27    57a29d44e2013f0c5a5da1df  501767d1cb022453170004dc         20.732   
28    57a29d44e2013f0c5a5da1e0  501767d1cb022453170004dc         18.239   
29    57a29d44e2013f0c5a5da1e1  501767d1cb022453170004dc         17.720   
...                        ...                       ...            ...   
4970  57a29a16296d6d43c8d34716  56ea2b6a63428b4ec3000063         85.600   
4971  57a29a16296d6d43c8d34717  56ea2b6a63428b4ec3000063        144.699   
4972  57a29a16296d6d43c8d34719  56ea2b6a63428b4ec3000063         48.846   
4973  57a2c989e2013f7ece94f31d  56ea2b6a63428b4ec3000063        198.392   
4974  57a2c989e2013f7ece94f31e  56ea2b6a63428b4ec3000063         41.769   
4975  57a2ca57e2013f32e8dccfa9  56ea2b6a63428b4ec3000063         21.259   
4976  57a2ed36e2013f64d45da932  56ea2b6a63428b4ec3000063         92.705   
4977  57a30262e2013f76f394f5cd  56ea2b6a63428b4ec3000063         41.953   
4978  57a3f708e2013f6195dccd17  56ea2b6a63428b4ec3000063        184.744   
4979  57a3f708e2013f6195dccd18  56ea2b6a63428b4ec3000063         28.273   
4980  57a3f708e2013f6195dccd19  56ea2b6a63428b4ec3000063         15.462   
4981  57a3f708e2013f6195dccd1a  56ea2b6a63428b4ec3000063         49.827   
4982  57a2a35a296d6d22f946f2ed  570b4757cb63b02e36000457        216.470   
4983  57a2b0ac296d6d54bec9b01a  570b4757cb63b02e36000457        102.864   
4984  57a2b67fe2013f26945da32b  570b4757cb63b02e36000457         72.253   
4985  57a2c5e6e2013f43f95d9e96  570b4757cb63b02e36000457         67.257   
4986  57a2c5e6e2013f43f95d9e97  570b4757cb63b02e36000457        103.240   
4987  57a2c6a9296d6d194e46e8c6  570b4757cb63b02e36000457         63.966   
4988  57a3f469296d6d3535d34042  570b4757cb63b02e36000457         46.409   
4989  57a41909296d6d35f6d345b3  570b4757cb63b02e36000457         64.304   
4990  57a29c07e2013f454f5da19a  570b47d0651bb1017e000335         65.734   
4991  57a297bde2013f2213dcd48e  573c2410651bb1ae

#### making matrix with correlation for all workers

In [13]:
workers_list = worker['worker_id'].unique()
temp_list=[]
worker_mat_binary = pd.DataFrame(columns = ['Worker Number','Worker_id','No of tasks','Correlation'])
for i,j in enumerate (workers_list):
    dataframe = worker[worker['worker_id']==str(j)]
    count = dataframe.count(axis=0)[0]
    core = dataframe['work_duration'].corr(dataframe['accuracy'])
    temp = pd.DataFrame([[i,j,count,core]], columns=['Worker Number','Worker_id','No of tasks','Correlation'])
    temp_list.append(temp)
worker_mat_binary = worker_mat_binary.append(temp_list)

C:\Users\Chinmay\Anaconda3\lib\site-packages\numpy\lib\function_base.py:2487: RuntimeWarning: Degrees of freedom <= 0 for slice
  warnings.warn("Degrees of freedom <= 0 for slice", RuntimeWarning)


In [14]:
worker_mat_binary.sort_values(by='No of tasks',ascending = False)

Worker Number                 Worker_id  No of tasks  Correlation
0            67.0  54896d76fc4e7e0f300000b4        659.0    -0.055394
0            51.0  523949ca3465040cdb0078ac        384.0    -0.105735
0            16.0  512dbf09c32206698e000518        271.0     0.309306
0            13.0  5119d07bed964935a800008b        244.0    -0.015082
0             2.0  50531712cb022432f3000f27        195.0    -0.200715
0             1.0  503b89aacb02245d320001a4        136.0    -0.022037
0            87.0  5678e1a8fc4e7e9e0b000086        133.0    -0.029428
0            33.0  51e60f42863a99e7d5025708        124.0    -0.204268
0            49.0  5232d065863a99092101733d        121.0     0.029540
0            72.0  566fc9c6fc4e7e1195000d55        114.0    -0.105204
0             0.0  501767d1cb022453170004dc        102.0    -0.055874
0            55.0  525e0c13807fcd982f03e0a0        101.0    -0.077497
0            81.0  5678c14efc4e7e8c1500003d         92.0     0.243286
0            38.0  5208c83db684b001ec000c79         91.0     0.005492
0            69.0  5668f692fc4e7e2ef60007c0         87.0     0.035119
0            60.0  52833d27863a9990e702e220         86.0    -0.163555
0             4.0  50b59592ed964941970005b6         85.0     0.152589
0             5.0  50b6078eed9649419f000758         81.0    -0.003941
0            41.0  521f2fa7863a99cb29000159         81.0    -0.057685
0            71.0  566fc951fc4e7e3aac0008a8         80.0     0.222596
0            62.0  52ae981f863a9953ae064312         78.0     0.064908
0            39.0  521ae408b684b0370203c22e         72.0    -0.143441
0            56.0  526f7299346504becc043687         69.0    -0.053875
0            36.0  51ff1ed1863a99a5d3002c40         69.0    -0.210102
0            65.0  52dc8819b684b02135020212         69.0     0.067254
0             3.0  5059b9b4cb02246c1200017d         63.0    -0.182366
0            82.0  5678c86cfc4e7e6e7e00004c         62.0    -0.064670
0             8.0  50ca9696cb0224456b0012ba         59.0    -0.300000
0            70.0  566fc7ddfc4e7e7cd7000cff         58.0     0.279083
0            25.0  518ca54f863a9902a40010db         57.0     0.047082
..            ...                       ...          ...          ...
0            59.0  527484feb684b0734e021055         12.0     0.042693
0            45.0  5229b417b684b08fa8006f8b         12.0     0.332587
0             9.0  50cdd243ed9649443f00035b         11.0    -0.488377
0            79.0  5677cd34fc4e7e267400002c         11.0    -0.440489
0            23.0  5178ebb6ed9649765a000166         11.0     0.305603
0            63.0  52bad53e863a999598061739         10.0    -0.284690
0            74.0  56711d03fc4e7edcd20009f4         10.0    -0.197534
0            92.0  573c2410651bb1ae90000516          9.0     0.306417
0            28.0  51a33eb1863a99725c00396c          9.0    -0.049449
0            90.0  570b4757cb63b02e36000457          8.0     0.598000
0            73.0  56711cfafc4e7edcd20009e6          8.0    -0.431036
0            61.0  52a6f377863a99a06f001cc4          8.0     0.934569
0            43.0  522847d0b684b0acd8003551          7.0    -0.205985
0            11.0  51089b31c322063455000002          6.0    -0.099010
0            68.0  54d95e9ffc4e7e009100281d          6.0    -0.369340
0            84.0  5678c87cfc4e7e6e7e000068          4.0     0.994158
0            42.0  5227f4bc863a996ea200ea7a          4.0    -0.508569
0            44.0  52286b2eb684b05f3f00616e          4.0     0.004472
0            19.0  513ff1fac3220622d8002e1c          4.0     0.136509
0            35.0  51e65c79b684b072b802aa05          2.0     1.000000
0            32.0  51c922b8863a99745f013211          2.0    -1.000000
0            40.0  521ae474b684b0a1a903cb63          2.0     1.000000
0            31.0  51c7f982b684b0ab8f026427          2.0    -1.000000
0            58.0  527345fb34650437bd011f58          1.0          NaN
0            14.0  512c57a2ed96497cdf0015e0          1.0          NaN
0            1

> If we see the above matrix, most of the people have negative correlation with work duration to accuracy. And that is very logical. 

In [15]:
worker_mat_binary_count = worker_mat_binary

In [16]:
worker_mat_binary_count.loc[worker_mat_binary_count['Correlation'] < 0, 'Correlation'] = -1
worker_mat_binary_count.loc[worker_mat_binary_count['Correlation'] >= 0, 'Correlation'] = 1

In [17]:
worker_mat_binary_count['Correlation'].value_counts()

-1.0    47
 1.0    39
Name: Correlation, dtype: int64

> Out of all people who have done more than 100 tasks (12 such people) 84% (10) people have negative correlation with accuracy to work duration. That means as work duration increases, accuracy decreases